jobs/ml_jobs/retrieval_vector/app/factory/similar_offer.py

```diff
@@ -0,0 +1,179 @@
+    ) -> Dict:
+        """
+        Performs the search using the vector for a given item.
+
+        Args:
+            model (DefaultClient): The model used for searching.
+            request_data (PredictionRequest): Request data containing search parameters.
+            vector (Document): The vector representation of the item.
+            item_id (str): The item ID for which the search is performed.
+
+        Returns:
+            Dict: A dictionary containing the search results.
+        """
+        return search_by_vector(
+            model=model,
+            vector=vector,
+            size=request_data.size
+            + 1,  # Add 1 to the size to account for the item that will be removed
+            selected_params=request_data.params,
+            debug=request_data.debug,
+            call_id=request_data.call_id,
+            prefilter=request_data.is_prefilter,
+            vector_column_name=request_data.vector_column_name,
+            similarity_metric=request_data.similarity_metric,
+            re_rank=request_data.re_rank,
+            item_id=item_id,
+            user_id=request_data.user_id,
+        )
+
+    def _select_best_predictions(
+        self, predictions: List[Dict], size: int, items: List[str]
+    ) -> List[Dict]:
+        """
+        Selects the best predictions by calculating the mean `_distance` value for each item,
+        and then sorting and returning the top X predictions based on the smallest mean `_distance`.
+
+        Args:
+            predictions (List[Dict]): A list of predictions.
+            size (int): The number of top results to return.
+
+        Returns:
+            List[Dict]: A list of predictions with the mean `_distance` for each item.
+        """
+        grouped_predictions = defaultdict(list)
+
+        # Group predictions by item_id
+        for entry in predictions:
+            if entry["item_id"] not in items:
+                grouped_predictions[entry["item_id"]].append(entry)
+
+        averaged_predictions = []
+        for item_id, entries in grouped_predictions.items():
+            mean_distance = self._calculate_mean_distance(entries)
+            # Use the first entry as a template, but update the distance to the mean distance
+            best_entry = entries[0].copy()
+            best_entry["_distance"] = mean_distance
+            averaged_predictions.append(best_entry)
+
+        averaged_predictions.sort(key=lambda x: x["_distance"])
+        return averaged_predictions[:size]
+
+    def _calculate_mean_distance(self, entries: List[Dict]) -> float:
+        """
+        Calculates the mean `_distance` for a list of predictions.
+
+        Args:
+            entries (List[Dict]): A list of prediction entries for the same item.
+
+        Returns:
+            float: The mean `_distance` value for the predictions.
+        """
+        total_distance = sum(entry["_distance"] for entry in entries)
+        return total_distance / len(entries)
```

Comment:
suggestion en oneline : 
`return np.mean([entry["_distance"] for entry in entries])`